### PR TITLE
Dashboard enhancements

### LINKED ISSUE #202: Dashboards enhancements
See excel -> https://docs.google.com/spreadsheets/d/1FRZz2RZm6Q6Zs7L0-c4ML0OoNwLrW_jM3KPqjeKu_G0/edit#gid=1504910404

- [x] Table: Adm + Coverage
**columns** doses + age group + sex + disp Status 
**rows** period
**one table per** antigen and site 
**Problem** how to calculate the coverage
**Solution** https://github.com/EyeSeeTea/vaccination-app/issues/215
http://dev2.eyeseetea.com:8082/dhis-web-pivot/?id=vPUwIbmBGYo

- [x] Table: Adm + Coverage
**columns** per doses + age group + sex + disp Status
**rows** area
**one table per** antigen
**Problem** how to calculate the coverage
**Solution** https://github.com/EyeSeeTea/vaccination-app/issues/215

- [x] Chart: Coverage
**columns** site
**dissagregated** doses  
**custom title** Coverage by site with start date XX and end date XX
**one table ** per antigen
**Problem** how to calculate the coverage
**Solution** https://github.com/EyeSeeTea/vaccination-app/issues/215

- [x] Chart: Coverage
**columns** are
**dissagregated** doses  
**custom title** Coverage by site with start date XX and end date XX
**one table** per antigen
**Problem** how to calculate the coverage
**Solution** https://github.com/EyeSeeTea/vaccination-app/issues/215

- [x] Table: Adm/Used/Utilization RAte
**columns** N/A
**rows** period and team
**one table per** antigen 
http://dev2.eyeseetea.com:8082/dhis-web-pivot/?id=UHlx3KqfnlH

- [x] Table: Adm/Used/Utilization RAte
**columns** N/A
**rows** area
**one table per** antigen 
http://dev2.eyeseetea.com:8082/dhis-web-pivot/?id=i3dpf1NZwqx

- [x] Table: Dilution Syringes Ratio and Needles Synringes Ratio
**columns** N/A
**rows** period and team
**one table per** antigen
**Problem** Safesty Boxes Ratio and ADS wastage rate

- [x] Table: Dilution Syringes Ratio, Needles Synringes Ratio, Safesty Boxes Ratio and ADS wastage rate
**columns** N/A
**rows** area
**one table per** campaign
http://dev2.eyeseetea.com:8082/dhis-web-pivot/?id=C70jIZepVOV

- [x] Table: AEFI & AEB
**columns** N/A
**rows** period
**one table per** campaign
**legend** greater than 0 -> background in red
http://dev2.eyeseetea.com:8082/dhis-web-pivot/?id=B3KUzXwbru4

### PR DESCRIPTION
**Pull request by matiasgarcia91**
### :pushpin: References

* **Issue:** Closes #202

### :memo: Implementation:
Created Tables and charts from ticket #202.
Refactored getMetadataForDashboardItems to use metadataConfig.

### :art: Screenshots
image

### :fire: Is there anything the reviewer should know to test it?
DHIS Metadata must include the legend creation https://github.com/EyeSeeTea/vaccination-metadata/pull/20.

### CODE REVIEW

File: src/models/dashboard-items.js [hunk 1]
```diff
@@ -85,44 +133,66 @@ function getDisaggregations(itemConfigs, disaggregationMetadata, antigen) {
...
             chartConstructor({
                 id: generateUid(),
                 antigen,
                 data: elements[key],
                 type: chart.type,
                 appendCode: chart.appendCode,
-                organisationUnit,
+                organisationUnits,
+                title: chart.title,
+                area: chart.area || false,
```
Comment (tokland) [inline src/models/dashboard-items.js]: This converts to boolean? Maybe the typical: `!!chart.area`

File: src/models/dashboard-items.js [hunk 2]
```diff
@@ -133,70 +203,115 @@ export function buildDashboardItems(
...
+        tables: globalTablesMetadata,
+        antigen: null,
+        elements,
+        organisationUnits: organisationUnitsMetadata,
+        itemsMetadata,
+        disaggregationMetadata,
+        legendsMetadata,
+    });
+
+    const reportTables = [
```
Comment (tokland) [inline src/models/dashboard-items.js]: We can use `_.concat(xs, ys, zs)`.

File: src/models/dashboard-items.js [hunk 3]
```diff
@@ -255,58 +370,80 @@ const chartConstructor = ({
...
+    filterDataBy,
+    title,
 }) => {
     const { categoryDimensions, columns: allColumns } = getDimensions(
         disaggregations,
         antigen,
         antigenCategory
     );
 
+    const periodForTitle = `${moment(period[0].id).format("DD/MM/YYYY")} - ${moment(
```
Comment (tokland) [inline src/models/dashboard-items.js]: It's a bit confusing, `period` (singular name) is an array.

File: src/models/dashboard-items.js [hunk 4]
```diff
@@ -255,58 +370,80 @@ const chartConstructor = ({
...
+        period[period.length - 1].id
```
Comment (tokland) [inline src/models/dashboard-items.js]: `_.last(period).id` ?

**Comment by tokland** [comment]
Let me add some screenshots about my campaign test. I don't see anything wrong, but you are more familiar with the requirements:


image

image

image

image

image

image

**Comment by adrianq** [comment]
@matiasgarcia91 I have done some initial testing and it looks like there is something wrong here. A few screenshots showing the data:
image
image

and now the dashboards:
Measles doesn't have any dose and it looks like there is not any value although there are some of them.
image

We are mixing teams from different campaigns:
image

Now other two different teams:
image

Also @tokland screenshot looks really weird. It is mixing old and new dashboard items. I will do some more testing but it actually looks like there is something wrong

**Comment by adrianq** [comment]
Ok, the error with Measles was not actually an error. I needed to re-save the population distribution values as they were not properly saved. The only minor thing is that it would be nice to not show 'Dose 1' when there is actually no doses. Not sure if it is easy to do at this point.

The teams bug, though, seem to be there

**Comment by adrianq** [comment]
@matiasgarcia91 Thanks for all the fixes. We are almost there. The bug team seems to be fixed on edit but not on creation. Yesterday, at first glance, I didn't realise but it looks like the error is still there. The main problem is what you detected teams are undefined and they get filled with whatever DHIS2 found.

@matiasgarcia91 @tokland whoever wakes up first give me a shout, I need a bit of help to fix this and probably it is gonna much faster with any of you by my side. Thanks!

**Pull request merged**